File: client/java/client/src/test/java/dev/angzarr/client/ErrorIntrospectionTest.java
package dev.angzarr.client;

import io.grpc.Status;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Tests for error introspection methods.
 *
 * Error introspection allows callers to check the nature of errors
 * without type casting or exception handling boilerplate.
 */
class ErrorIntrospectionTest {

    // =========================================================================
    // isNotFound Tests
    // =========================================================================

    @Test
    void grpcError_with_NOT_FOUND_should_return_true_for_isNotFound() {
        var error = new Errors.GrpcError("not found", Status.Code.NOT_FOUND);
        assertThat(error.isNotFound()).isTrue();
    }

    @Test
    void grpcError_with_other_code_should_return_false_for_isNotFound() {
        var error = new Errors.GrpcError("internal error", Status.Code.INTERNAL);
        assertThat(error.isNotFound()).isFalse();
    }

    // =========================================================================
    // isPreconditionFailed Tests
    // =========================================================================

    @Test
    void grpcError_with_FAILED_PRECONDITION_should_return_true_for_isPreconditionFailed() {
        var error = new Errors.GrpcError("precondition failed", Status.Code.FAILED_PRECONDITION);
        assertThat(error.isPreconditionFailed()).isTrue();
    }

    @Test
    void grpcError_with_other_code_should_return_false_for_isPreconditionFailed() {
        var error = new Errors.GrpcError("internal error", Status.Code.INTERNAL);
        assertThat(error.isPreconditionFailed()).isFalse();
    }

    @Test
    void commandRejectedError_should_return_true_for_isPreconditionFailed() {
        // CommandRejectedError defaults to FAILED_PRECONDITION
        var error = new Errors.CommandRejectedError("rejected");
        assertThat(error.isPreconditionFailed()).isTrue();
    }

    // =========================================================================
    // isInvalidArgument Tests
    // =========================================================================

    @Test
    void grpcError_with_INVALID_ARGUMENT_should_return_true_for_isInvalidArgument() {
        var error = new Errors.GrpcError("invalid argument", Status.Code.INVALID_ARGUMENT);
        assertThat(error.isInvalidArgument()).isTrue();
    }

    @Test
    void invalidArgumentError_should_return_true_for_isInvalidArgument() {
        var error = new Errors.InvalidArgumentError("bad input");
        assertThat(error.isInvalidArgument()).isTrue();
    }

    @Test
    void grpcError_with_other_code_should_return_false_for_isInvalidArgument() {
        var error = new Errors.GrpcError("internal error", Status.Code.INTERNAL);
        assertThat(error.isInvalidArgument()).isFalse();
    }

    // =========================================================================
    // isConnectionError Tests
    // =========================================================================

    @Test
    void connectionError_should_return_true_for_isConnectionError() {
        var error = new Errors.ConnectionError("connection refused");
        assertThat(error.isConnectionError()).isTrue();
    }

    @Test
    void transportError_should_return_true_for_isConnectionError() {
        var error = new Errors.TransportError("transport failed");
        assertThat(error.isConnectionError()).isTrue();
    }

    @Test
    void grpcError_with_UNAVAILABLE_should_return_true_for_isConnectionError() {
        var error = new Errors.GrpcError("unavailable", Status.Code.UNAVAILABLE);
        assertThat(error.isConnectionError()).isTrue();
    }

    @Test
    void grpcError_with_other_code_should_return_false_for_isConnectionError() {
        var error = new Errors.GrpcError("internal error", Status.Code.INTERNAL);
        assertThat(error.isConnectionError()).isFalse();
    }

    // =========================================================================
    // Base Class Default Behavior Tests
    // =========================================================================

    @Test
    void clientError_should_have_default_false_for_all_introspection_methods() {
        // Direct ClientError instance should return false for all introspection
        var error = new Errors.ClientError("generic error");
        assertThat(error.isNotFound()).isFalse();
        assertThat(error.isPreconditionFailed()).isFalse();
        assertThat(error.isInvalidArgument()).isFalse();
        assertThat(error.isConnectionError()).isFalse();
    }
}

File: client/cpp/tests/error_introspection_test.cpp
#include <gtest/gtest.h>
#include "angzarr/errors.hpp"

using namespace angzarr;

// =============================================================================
// IsNotFound Tests
// =============================================================================

TEST(ErrorIntrospectionTest, GrpcError_WithNotFound_ShouldReturnTrueForIsNotFound) {
    GrpcError error("not found", grpc::StatusCode::NOT_FOUND);
    EXPECT_TRUE(error.is_not_found());
}

TEST(ErrorIntrospectionTest, GrpcError_WithOtherCode_ShouldReturnFalseForIsNotFound) {
    GrpcError error("internal error", grpc::StatusCode::INTERNAL);
    EXPECT_FALSE(error.is_not_found());
}

// =============================================================================
// IsPreconditionFailed Tests
// =============================================================================

TEST(ErrorIntrospectionTest, GrpcError_WithFailedPrecondition_ShouldReturnTrueForIsPreconditionFailed) {
    GrpcError error("precondition failed", grpc::StatusCode::FAILED_PRECONDITION);
    EXPECT_TRUE(error.is_precondition_failed());
}

TEST(ErrorIntrospectionTest, GrpcError_WithOtherCode_ShouldReturnFalseForIsPreconditionFailed) {
    GrpcError error("internal error", grpc::StatusCode::INTERNAL);
    EXPECT_FALSE(error.is_precondition_failed());
}

TEST(ErrorIntrospectionTest, CommandRejectedError_ShouldReturnTrueForIsPreconditionFailed) {
    CommandRejectedError error("rejected");
    EXPECT_TRUE(error.is_precondition_failed());
}

// =============================================================================
// IsInvalidArgument Tests
// =============================================================================

TEST(ErrorIntrospectionTest, GrpcError_WithInvalidArgument_ShouldReturnTrueForIsInvalidArgument) {
    GrpcError error("invalid argument", grpc::StatusCode::INVALID_ARGUMENT);
    EXPECT_TRUE(error.is_invalid_argument());
}

TEST(ErrorIntrospectionTest, InvalidArgumentError_ShouldReturnTrueForIsInvalidArgument) {
    InvalidArgumentError error("bad input");
    EXPECT_TRUE(error.is_invalid_argument());
}

TEST(ErrorIntrospectionTest, GrpcError_WithOtherCode_ShouldReturnFalseForIsInvalidArgument) {
    GrpcError error("internal error", grpc::StatusCode::INTERNAL);
    EXPECT_FALSE(error.is_invalid_argument());
}

// =============================================================================
// IsConnectionError Tests
// =============================================================================

TEST(ErrorIntrospectionTest, ConnectionError_ShouldReturnTrueForIsConnectionError) {
    ConnectionError error("connection refused");
    EXPECT_TRUE(error.is_connection_error());
}

TEST(ErrorIntrospectionTest, TransportError_ShouldReturnTrueForIsConnectionError) {
    TransportError error("transport failed");
    EXPECT_TRUE(error.is_connection_error());
}

TEST(ErrorIntrospectionTest, GrpcError_WithUnavailable_ShouldReturnTrueForIsConnectionError) {
    GrpcError error("unavailable", grpc::StatusCode::UNAVAILABLE);
    EXPECT_TRUE(error.is_connection_error());
}

TEST(ErrorIntrospectionTest, GrpcError_WithOtherCode_ShouldReturnFalseForIsConnectionError) {
    GrpcError error("internal error", grpc::StatusCode::INTERNAL);
    EXPECT_FALSE(error.is_connection_error());
}

// =============================================================================
// Base Class Default Behavior Tests
// =============================================================================

TEST(ErrorIntrospectionTest, ClientError_ShouldHaveDefaultFalseForAllIntrospectionMethods) {
    ClientError error("generic error");
    EXPECT_FALSE(error.is_not_found());
    EXPECT_FALSE(error.is_precondition_failed());
    EXPECT_FALSE(error.is_invalid_argument());
    EXPECT_FALSE(error.is_connection_error());
}
